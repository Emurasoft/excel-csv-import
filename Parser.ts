/* global Office */
import * as ExcelAPI from './ExcelAPI';
import * as Papa from 'papaparse';
import {AbortFlag} from './AbortFlag';
import {ProgressCallback} from './Store';
import {APIVersionInfo, Shape} from './ExcelAPI';

export const enum InputType {file, text}

export interface Source {
	inputType: InputType;
	file?: File;
	text: string;
}

type Config = {
	[P in 'delimiter' | 'encoding']: Papa.ParseConfig[P];
}

export const enum NewlineSequence {
	AutoDetect = '',
	CRLF = '\r\n',
	CR = '\r',
	LF = '\n',
}

export interface ImportOptions extends Config {
	source: Source;
	newline: NewlineSequence;
}

export interface ExportOptions {
	delimiter: string;
	newline: NewlineSequence;
}

let reduceChunkSize = null;

export async function init(): Promise<APIVersionInfo> {
	const result = await ExcelAPI.init();
	if (result.platform === Office.PlatformType.OfficeOnline) {
		// Online API can throw error if request size is too large
		reduceChunkSize = true;
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		(Papa.LocalChunkSize as any) = 10_000;
	} else {
		reduceChunkSize = false;
	}
	return result;
}

export class ChunkProcessor {
	public constructor(
		worksheet: Excel.Worksheet,
		progressCallback: ProgressCallback,
		abortFlag: AbortFlag,
	) {
		this._worksheet = worksheet;
		this._progressCallback = progressCallback;
		this._abortFlag = abortFlag;
		this._currRow = 0;
		this._currentProgress = 0.0;
	}

	public run(importOptions: ImportOptions & Papa.ParseConfig): Promise<Papa.ParseError[]> {
		this._progressCallback(0.0);
		this._progressPerChunk = ChunkProcessor.progressPerChunk(
			importOptions.source,
			// @ts-ignore
			Papa.LocalChunkSize,
		);

		return new Promise(resolve => {
			importOptions.chunk = this.chunk;
			importOptions.complete = results => resolve(results.errors);

			switch (importOptions.source.inputType) {
				case InputType.file:
					Papa.parse(importOptions.source.file, importOptions);
					break;
				case InputType.text:
					Papa.parse(importOptions.source.text, importOptions);
					break;
			}
		});
	}

	private static progressPerChunk(source: Source, chunkSize: number): number {
		switch (source.inputType) {
			case InputType.file:
				if (source.file.size === 0) {
					return 1.0;
				}
				return chunkSize / source.file.size;
			case InputType.text:
				if (source.text.length === 0) {
					return 1.0;
				}
				return chunkSize / source.text.length;
		}
	}

	private readonly _worksheet: Excel.Worksheet;
	private readonly _progressCallback: ProgressCallback;
	private readonly _abortFlag: AbortFlag;
	private readonly _excelAPI = ExcelAPI;
	private _currRow: number;
	private _progressPerChunk: number;
	private _currentProgress: number;

	private chunk = (chunk: Papa.ParseResult, parser: Papa.Parser) => {
		if (this._abortFlag.aborted()) {
			parser.abort();
		}

		this._worksheet.context.application.suspendApiCalculationUntilNextSync();
		this._excelAPI.setChunk(this._worksheet, this._currRow, chunk.data);
		this._currRow += chunk.data.length;
		parser.pause();
		// sync() must be called after each chunk, otherwise API may throw exception
		this._worksheet.context.sync().then(parser.resume);
		// Since the Excel API is so damn slow, updating GUI every chunk has a negligible impact
		// on performance.
		this._progressCallback(this._currentProgress += this._progressPerChunk);
	}
}

export async function importCSV(
	importOptions: ImportOptions,
	progressCallback: ProgressCallback,
	abortFlag: AbortFlag,
): Promise<Papa.ParseError[]> {
	let errors = null;
	await ExcelAPI.runOnBlankWorksheet(async (worksheet) => {
		const chunkProcessor = new ChunkProcessor(worksheet, progressCallback, abortFlag);
		errors = await chunkProcessor.run(importOptions);
	});
	return errors;
}

/*
RFC 4180 standard:
MS-DOS-style lines that end with (CR/LF) characters (optional for the last line).
An optional header record (there is no sure way to detect whether it is present, so care is required
when importing).
Each record "should" contain the same number of comma-separated fields.
Any field may be quoted (with double quotes).
Fields containing a line-break, double-quote or commas (delimiter) should be quoted. (If they are
not, the file will likely be impossible to process correctly).
A (double) quote character in a field must be represented by two (double) quote characters.
Thanks Wikipedia.
 */

/* eslint-disable @typescript-eslint/no-explicit-any */
export function _chunkRange(
	chunk: number,
	shape: Shape,
	chunkRows: number,
): {startRow: number; startColumn: number; rowCount: number; columnCount: number} {
	return {
		startRow: chunk * chunkRows,
		startColumn: 0,
		rowCount: Math.min(chunkRows, shape.rows),
		columnCount: shape.columns,
	};
}

export function _addQuotes(row: string[], delimiter: string): void {
	console.assert(delimiter !== '');
	const charactersToWatchOutFor = ['\r', '\n', '\u0022' /*double quote*/, delimiter];
	for (let i = 0; i < row.length; i++) {
		if (charactersToWatchOutFor.some(c => row[i].includes(c))) {
			row[i] = '\u0022' + row[i].replace(/\u0022/g, '\u0022\u0022') + '\u0022';
		}
	}
}

export function _rowString(row: any[], exportOptions: Readonly<ExportOptions>): string {
	const stringValues = row.map(a => a.toString());
	_addQuotes(stringValues, exportOptions.delimiter);
	return stringValues.join(exportOptions.delimiter) + exportOptions.newline;
}

export function _chunkString(values: any[][], exportOptions: Readonly<ExportOptions>): string {
	let result = '';

	for (let i = 0; i < values.length; i++) {
		result += _rowString(values[i], exportOptions);
	}

	return result;
}

export async function _csvString(
	worksheet: Excel.Worksheet,
	shape: Shape,
	chunkRows: number,
	exportOptions: Readonly<ExportOptions>,
	progressCallback: ProgressCallback,
	abortFlag: AbortFlag,
): Promise<string> {
	let result = '';

	// chunkRows is never 0
	for (let chunk = 0; chunk < Math.ceil(shape.rows / chunkRows); chunk++) {
		if (abortFlag.aborted()) {
			break;
		}

		// shape.rows is never 0
		progressCallback(chunk * chunkRows / shape.rows);

		const chunkRange = _chunkRange(chunk, shape, chunkRows);
		const range = worksheet.getRangeByIndexes(
			chunkRange.startRow,
			chunkRange.startColumn,
			chunkRange.rowCount,
			chunkRange.columnCount,
		).load('values');
		await worksheet.context.sync();

		result += _chunkString(range.values, exportOptions);
	}

	return result;
}
/* eslint-enable @typescript-eslint/no-explicit-any */

export function _nameToUse(workbookName: string, worksheetName: string): string {
	if (/^Sheet\d+$/.test(worksheetName)) { // 'Sheet1' isn't a good name to use
		// Workbook name usually includes the file extension
		const to = workbookName.lastIndexOf('.');
		return workbookName.substr(0, to === -1 ? workbookName.length : to);
	} else {
		return worksheetName;
	}
}

function chunkRows(shape: Shape): number {
	if (reduceChunkSize) {
		return Math.floor(10_000 / shape.columns)
	} else {
		return shape.rows;
	}
}

export interface CsvStringAndName {
	name: string;
	string: string;
}

export async function csvStringAndName(
	exportOptions: ExportOptions,
	progressCallback: ProgressCallback,
	abortFlag: AbortFlag,
): Promise<CsvStringAndName> {
	let namesAndShape = null;
	let resultString = '';
	await ExcelAPI.runOnCurrentWorksheet(async (worksheet) => {
		namesAndShape = await ExcelAPI.worksheetNamesAndShape(worksheet);
		worksheet.context.application.suspendApiCalculationUntilNextSync();
		resultString = await _csvString(
			worksheet,
			namesAndShape.shape,
			chunkRows(namesAndShape.shape),
			exportOptions,
			progressCallback,
			abortFlag,
		);
	});

	return {
		name: _nameToUse(namesAndShape.workbookName, namesAndShape.worksheetName),
		string: resultString,
	};
}