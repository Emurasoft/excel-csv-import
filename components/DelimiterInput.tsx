import * as React from 'react';
import {Body1, Dropdown, Input, Label, Option, Subtitle1} from '@fluentui/react-components';
import {useState} from 'react';
import { useStyles } from './styles';

export const enum DropdownOption {
	comma = 'Comma',
	space = 'Space',
	tab = 'Tab',
	other = 'Other',
}

interface Props {
	showLengthError: boolean;
	value: string;
	onChange: (value: string) => void;
}

const stringToDropdownKey = {
	'\u002c': DropdownOption.comma,
	'\u0020': DropdownOption.space,
	'\u0009': DropdownOption.tab,
};

const dropdownToString = {
	[DropdownOption.comma]: '\u002c',
	[DropdownOption.space]: '\u0020',
	[DropdownOption.tab]: '\u0009',
	[DropdownOption.other]: '',
};

export function DelimiterInput({showLengthError, value, onChange}: Props): React.ReactElement {
	const [otherSelected, setOtherSelected] = useState(false);
	const styles = useStyles();

	const selectedKey = (): DropdownOption => {
		if (!otherSelected && value in stringToDropdownKey) {
			return stringToDropdownKey[value];
		}

		return DropdownOption.other;
	}

	const customInput = (
		<div className="smallDivider">
			<Input
				className={styles.monospace}
				value={value}
				onChange={(_, {value}) => onChange(value)}
				placeholder="Enter custom delimiter"
				spellCheck={false}
			/>
			<br />
			<Body1>{description(value)}</Body1>
			{
				showLengthError && value.length > 1 
				&& <Body1 className={styles.redText}>
					Delimiter length must be 1
				</Body1>
			}
		</div>
	);

	return (
		<>
			<Label> 
				<Subtitle1>Delimiter</Subtitle1>
				<br />
				<Dropdown
					placeholder="Delimiter"
					value={selectedKey()}
					onOptionSelect={(_, {optionValue}) => {
						setOtherSelected(optionValue === DropdownOption.other);
						onChange(dropdownToString[optionValue]);
					}}
				>
					<Option>{DropdownOption.comma}</Option>
					<Option>{DropdownOption.space}</Option>
					<Option>{DropdownOption.tab}</Option>
					<Option>{DropdownOption.other}</Option>
				</Dropdown>
				{
					otherSelected || !['\u002c', '\u0020', '\u0009'].includes(value)
						? customInput
						: null
				}
			</Label>
		</>
	);
}

function description(delimiter: string): string {
	if (delimiter.length == 1) {
		return codePoint(delimiter);
	} else {
		return '\u00A0';
	}
}

export function codePoint(c: string): string {
	return 'U+' + c[0].charCodeAt(0).toString(16).toUpperCase().padStart(4, '0');
}