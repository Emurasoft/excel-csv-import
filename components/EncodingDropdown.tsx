import * as React from 'react';
import {Dropdown, Label, Option, Subtitle1} from '@fluentui/react-components';
import {EncodingDropdownOptions} from './EncodingDropdownOptions';

interface Props {
	showAutoDetect: boolean;
	value: string;
	onChange: (value: string) => void;
}

export function EncodingDropdown({showAutoDetect, value, onChange}: Props): React.ReactElement {
	let dropdownOptions: string[];
	if (showAutoDetect) {
		dropdownOptions = ['Auto-detect', ...EncodingDropdownOptions];
	} else {
		dropdownOptions = EncodingDropdownOptions;
	}

	return (
		<Label>
			<Subtitle1>Encoding</Subtitle1>
			<br />
			<Dropdown
				value={value}
				onOptionSelect={(_, {optionValue}) => onChange(optionValue)}
			>
				{dropdownOptions.map((v) => <Option key={v}>{v}</Option>)}
			</Dropdown>
		</Label>
	);
}