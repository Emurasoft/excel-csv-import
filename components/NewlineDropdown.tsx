import * as React from 'react';
import {Dropdown, Label, Option, Subtitle1} from '@fluentui/react-components';
import {NewlineSequence} from '../parser';

interface Props {
	showAutoDetect: boolean;
	value: NewlineSequence;
	onChange: (value: NewlineSequence) => void;
}

function getLabel(n: NewlineSequence): string {
	switch (n) {
	case NewlineSequence.AutoDetect:
		return 'Auto-detect';
	case NewlineSequence.CRLF:
		return 'CRLF';
	case NewlineSequence.CR:
		return 'CR';
	case NewlineSequence.LF:
		return 'LF';
	}
}

export function NewlineDropdown({showAutoDetect, value, onChange}: Props): React.ReactElement {
	return (
		<Label>
			<Subtitle1>Newline sequence</Subtitle1>
			<br />
			<Dropdown
				value={getLabel(value)}
				onOptionSelect={(_, {optionValue}) => onChange(optionValue as NewlineSequence)}
			>
				{
					showAutoDetect
					&& <Option value={NewlineSequence.AutoDetect}>
						{getLabel(NewlineSequence.AutoDetect)}
					</Option>
				}
				<Option value={NewlineSequence.CRLF}>
					{getLabel(NewlineSequence.CRLF)}
				</Option>
				<Option value={NewlineSequence.CR}>
					{getLabel(NewlineSequence.CR)}
				</Option>
				<Option value={NewlineSequence.LF}>
					{getLabel(NewlineSequence.LF)}
				</Option>
			</Dropdown>
		</Label>
	);
}