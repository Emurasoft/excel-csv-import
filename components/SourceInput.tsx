import * as React from 'react';
import { InputType, Source } from '../parser';
import { Dropdown, Label, mergeClasses, Option, Subtitle1, Textarea } from '@fluentui/react-components';
import { useStyles } from './styles';

interface Props {
	value: Source;
	onChange: (value: Source) => void;
}

function fileInput(onChange: (value: File) => void): React.ReactElement {
	const styles = useStyles();
	return (
		<>
			<input
				className={styles.fullWidth}
				type="file"
				accept="text/csv"
				onChange={e => onChange(e.target.files[0])}
				id="SourceInput-FileInput"
			/>
			<br />
		</>
	);
}

function textInput(value: string, onChange: (value: string) => void): React.ReactElement {
	const styles = useStyles();
	return (
		<Textarea
			className={mergeClasses(styles.monospace, styles.fullWidth)}
			rows={10}
			spellCheck={false}
			wrap="off"
			onChange={(_, { value }) => onChange(value)}
			value={value}
			id="SourceInput-TextInput"
		/>
	);
}

export function SourceInput({ value, onChange }: Props): React.ReactElement {
	let input: React.ReactElement;
	switch (value.inputType) {
		case InputType.file:
			input = fileInput(file => onChange({ inputType: InputType.file, file, text: '' }));
			break;
		case InputType.text:
			input = textInput(
				value.text,
				text => onChange({ inputType: InputType.text, file: null, text }),
			);
	}

	return (
		<>
			<Label>
				<Subtitle1>Import type</Subtitle1>
				<br />
				<Dropdown
					value={value.inputType}
					onOptionSelect={(_, { optionValue }) => onChange(
						{ inputType: optionValue as InputType, file: null, text: '' },
					)}
					id="SourceInput-Dropdown"
				>
					<Option>
						{InputType.file}
					</Option>
					<Option>
						{InputType.text}
					</Option>
				</Dropdown>
			</Label>
			{input}
		</>
	);
}
