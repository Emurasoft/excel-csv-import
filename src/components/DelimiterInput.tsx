import * as React from 'react';
import {Dropdown, IDropdownOption, TextField, ResponsiveMode} from '@fluentui/react';
import * as style from './style.css';
import {BaseProps} from './BaseProps';

export enum DropdownOptionKey {comma, space, tab, other}

interface Props extends BaseProps<string> {
    showLengthError: boolean;
}

interface State {
    otherSelected: boolean;
}

export class DelimiterInput extends React.Component<Props, State> {
    public constructor(props) {
        super(props);
        this.state = {otherSelected: false};

        this._stringToDropdownKey = {
            '\u002c': DropdownOptionKey.comma,
            '\u0020': DropdownOptionKey.space,
            '\u0009': DropdownOptionKey.tab,
        };
    }

    public render(): React.ReactNode {

        const dropdownOptions: IDropdownOption[] = [
            {
                key: DropdownOptionKey.comma,
                text: 'Comma (U+002C)',
            },
            {
                key: DropdownOptionKey.space,
                text: 'Space (U+0020)',
            },
            {
                key: DropdownOptionKey.tab,
                text: 'Tab (U+0009)',
            },
            {
                key: DropdownOptionKey.other,
                text: 'Other',
            },
        ];

        const customInput = (
            <div className={style.smallDivider}>
                <TextField
                    className={style.monospace}
                    value={this.props.value}
                    onChange={(_, value) => this.props.onChange(value)}
                    description={DelimiterInput.description(this.props.value)}
                    onGetErrorMessage={this.getErrorMessage}
                    deferredValidationTime={1}
                    placeholder={'Enter custom delimiter'}
                    spellCheck={false}
                />
            </div>
        );

        return (
            <>
                <Dropdown
                    label={'Delimiter'}
                    options={dropdownOptions}
                    responsiveMode={ResponsiveMode.large}
                    selectedKey={this.selectedKey()}
                    onChange={this.dropdownOnChange}
                />
                {this.showCustomInput() ? customInput : null}
            </>
        );
    }

    private static description(delimiter: string): string {
        if (delimiter.length == 1) {
            return DelimiterInput.codePoint(delimiter);
        } else {
            return '\u00A0';
        }
    }

    private static codePoint(c: string): string {
        return 'U+' + c[0].charCodeAt(0).toString(16).toUpperCase().padStart(4, '0');
    }

    private readonly _stringToDropdownKey: {[key: string]: DropdownOptionKey};

    private showCustomInput(): boolean {
        if (this.state.otherSelected) {
            return true;
        }

        return !['\u002c', '\u0020', '\u0009'].includes(this.props.value);
    }

    private selectedKey(): DropdownOptionKey {
        if (this.state.otherSelected) {
            return DropdownOptionKey.other;
        }

        if (this.props.value in this._stringToDropdownKey) {
            return this._stringToDropdownKey[this.props.value];
        } else {
            return DropdownOptionKey.other;
        }
    }

    private dropdownOnChange = (_, option: IDropdownOption) => {
        const dropdownToString = {
            [DropdownOptionKey.comma]: '\u002c',
            [DropdownOptionKey.space]: '\u0020',
            [DropdownOptionKey.tab]: '\u0009',
            [DropdownOptionKey.other]: '',
        };

        this.setState({otherSelected: option.key === DropdownOptionKey.other});
        this.props.onChange(dropdownToString[option.key]);
    }

    private getErrorMessage = (value: string) => {
        if (this.props.showLengthError && value.length > 1) {
            return 'Delimiter length must be 1';
        } else {
            return '';
        }
    }
}
