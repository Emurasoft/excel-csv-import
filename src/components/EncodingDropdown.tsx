import {BaseProps} from './BaseProps';
import * as React from 'react';
import {Dropdown, IDropdownOption} from 'office-ui-fabric-react';
import {EncodingDropdownOptions} from './EncodingDropdownOptions';
import {withTranslation} from 'react-i18next';

interface Props extends BaseProps<string> {
    showAutoDetect: boolean;
    hidden: boolean;
}

export class EncodingDropdownComponent extends React.Component<Props, {}> {
    public constructor(props: Props) {
        super(props);

        const AutoDetectOption: IDropdownOption = {
            'key': '',
            'text': props.t('Auto-detect'),
        };

        if (props.showAutoDetect) {
            this._dropdownOptions = [AutoDetectOption, ...EncodingDropdownOptions];
        } else {
            this._dropdownOptions = EncodingDropdownOptions;
        }
    }

    public render(): React.ReactNode {
        const t = this.props.t;
        if (this.props.hidden) {
            return null;
        } else {
            return (
                <>
                    <Dropdown
                        label={t('EncodingDropdown.Encoding')}
                        selectedKey={this.props.value}
                        options={this._dropdownOptions}
                        onChange={(_, option) => this.props.onChange(option.key as string)}
                    />
                    <br/>
                </>
            );
        }
    }

    private readonly _dropdownOptions: IDropdownOption[];
}

// @ts-ignore
export const EncodingDropdown = withTranslation('importExport')(EncodingDropdownComponent);