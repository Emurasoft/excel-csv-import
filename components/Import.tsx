import {Store} from '../Store';
import * as React from 'react';
import {connect} from '../connect';
import {PrimaryButton, Toggle, TooltipDelay, TooltipHost} from '@fluentui/react';
import {ImportOptions, InputType, NewlineSequence} from '../Parser';
import {SourceInput} from './SourceInput';
import {DelimiterInput} from './DelimiterInput';
import {NewlineDropdown} from './NewlineDropdown';
import {EncodingDropdown} from './EncodingDropdown';
import {ProgressBar} from './ProgressBar';
import * as style from './style.css';
import {BottomBar} from './BottomBar';
import {ParserOutputBox} from './ParserOutputBox';
import {StoredComponent} from './StoredComponent';
import {MemoryHistory} from 'history';
import {TitleBar} from './TitleBar';

interface Props {
    store: Store;
    history: MemoryHistory;
}

export class ImportComponent extends StoredComponent<Props, ImportOptions> {
    public constructor(props: Props) {
        super(props, 'import', {
            source: {inputType: InputType.file, file: null, text: ''},
            delimiter: '\u002c',
            newline: NewlineSequence.AutoDetect,
            encoding: '',
        }, ['delimiter', 'newline', 'encoding']);
    }

    public render(): React.ReactNode {
        return (
            <>
                <div className={style.pageMargin}>
                    {/* eslint-disable no-undef */}
                    <TitleBar
                        text={'Import CSV'}
                        helpLink={
                            'https://github.com/Emurasoft/excel-csv-import-help/blob/master/en.md'
                        }
                        mac={this.props.store.state.platform === Office.PlatformType.Mac}
                    />
                    {/* eslint-enable no-undef */}
                    <SourceInput
                        value={this.state.source}
                        onChange={(source) => this.setState({source})}
                    />
                    <br />
                    <EncodingDropdown
                        value={this.state.encoding}
                        onChange={(encoding) => this.setState({encoding})}
                        hidden={this.state.source.inputType === InputType.text}
                        showAutoDetect={true}
                    />
                    <DelimiterInput
                        value={this.state.delimiter}
                        onChange={(delimiter) => this.setState({delimiter})}
                        showLengthError={true}
                    />
                    <br />
                    <NewlineDropdown
                        value={this.state.newline as NewlineSequence}
                        onChange={(newline) => this.setState({newline})}
                        showAutoDetect={true}
                    />
                    <br />
                    <TooltipHost
                        styles={{root: {display: 'inline-block'}} /* Resize to fit button */}
                        content={this.buttonTooltipContent()}
                        delay={TooltipDelay.zero}
                    >
                        <PrimaryButton
                            disabled={this.buttonTooltipContent() !== ''}
                            onClick={this.buttonOnClick}
                            text={'Import CSV'}
                        />
                    </TooltipHost>
                    <br />
                    <ProgressBar
                        onClick={this.props.store.abort}
                        progress={this.props.store.state.progress}
                    />
                    <Toggle
                        inlineLabel label={'Save options'}
                        defaultChecked={this.initialSaveStatus()}
                        onChange={(_, checked) => this.setSaveStatus(checked)}
                    />
                    <ParserOutputBox parserOutput={this.props.store.state.parserOutput} />
                    <BottomBar />
                </div>
            </>
        );
    }

    private buttonOnClick = async () => {
        await this.props.store.import(this.state);
    }

    private buttonTooltipContent(): string {
        if (this.state.source.inputType == InputType.file && this.state.source.file == null) {
            return 'Import source is not selected';
        } else if (this.state.delimiter.length !== 1) {
            return 'Delimiter is invalid';
        } else if (!this.props.store.state.initialized) {
            return 'Excel API is not initialized';
        } else {
            return '';
        }
    }
}

export default connect(ImportComponent);