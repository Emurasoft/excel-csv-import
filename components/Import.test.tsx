import {ImportComponent} from './Import';
import {shallow} from 'enzyme';
import * as React from 'react';
import {InputType, NewlineSequence} from '../Parser';
import {SourceInput} from './SourceInput';
import {DelimiterInput} from './DelimiterInput';
import {NewlineDropdown} from './NewlineDropdown';
import {PrimaryButton} from '@fluentui/react';
import * as assert from 'assert';
import {EncodingDropdown} from './EncodingDropdown';

describe('ImportComponent', () => {
	afterEach(() => localStorage.clear());

	it('import', () => {
		let receivedOptions = null;
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		const stub: any = {};
		stub.state = {initialized: true};
		stub.import = (options) => receivedOptions = options
		// @ts-ignore
		const wrapper = shallow(<ImportComponent store={stub} />);

		wrapper.find(SourceInput)
			.simulate('change', {inputType: InputType.text, text: 'csv text'});
		wrapper.find(DelimiterInput).simulate('change', ',');
		wrapper.find(NewlineDropdown).simulate('change', NewlineSequence.LF);
		wrapper.find(EncodingDropdown).simulate('change', 'UTF-8');
		wrapper.find(PrimaryButton).simulate('click');

		const expected = {
			source: {inputType: 1, text: 'csv text'},
			delimiter: ',',
			newline: NewlineSequence.LF,
			encoding: 'UTF-8',
		};
		assert.deepStrictEqual(receivedOptions, expected);
	});
});