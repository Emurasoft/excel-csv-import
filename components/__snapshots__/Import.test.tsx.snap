// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import import 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      style="width: 100%; display: flex; justify-content: space-between;"
    >
      <span
        class="fui-Title1 fui-Text"
      >
        Import CSV
      </span>
      <a
        aria-label="Help page"
        class="fui-Link"
        href="https://github.com/Emurasoft/excel-csv-import-help/blob/master/en.md"
        rel="noopener noreferrer"
        tabindex="0"
        target="_blank"
      >
        <button
          class="fui-Button"
          style="margin-right: 4px;"
          type="button"
        >
          <span
            class="fui-Button__icon"
          >
            <svg
              aria-hidden="true"
              class=""
              fill="currentColor"
              height="32"
              viewBox="0 0 32 32"
              width="32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11 11a5 5 0 0 1 10 0c0 1.46-.33 2.31-.76 2.92-.41.6-.95 1.01-1.63 1.55l-.23.18a8.56 8.56 0 0 0-2.31 2.52A9.54 9.54 0 0 0 15 23a1 1 0 1 0 2 0c0-1.87.35-3.04.8-3.84a6.62 6.62 0 0 1 1.82-1.94l.23-.18c.66-.52 1.44-1.12 2.03-1.97A6.8 6.8 0 0 0 23 11a7 7 0 1 0-14 0 1 1 0 1 0 2 0Zm5 17.75a1.25 1.25 0 1 0 0-2.5 1.25 1.25 0 0 0 0 2.5Z"
                fill="currentColor"
              />
            </svg>
          </span>
        </button>
      </a>
    </div>
    <br />
    <label
      class="fui-Label"
    >
      <span
        class="fui-Subtitle1 fui-Text"
      >
        Import type
      </span>
      <br />
      <div
        class="fui-Dropdown"
      >
        <button
          aria-expanded="false"
          class="fui-Dropdown__button"
          id="SourceInput-Dropdown"
          role="combobox"
          tabindex="0"
          type="button"
          value="Text input"
        >
          Text input
          <span
            class="fui-Dropdown__expandIcon"
          >
            <svg
              aria-hidden="true"
              class=""
              fill="currentColor"
              height="1em"
              viewBox="0 0 20 20"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.85 7.65c.2.2.2.5 0 .7l-5.46 5.49a.55.55 0 0 1-.78 0L4.15 8.35a.5.5 0 1 1 .7-.7L10 12.8l5.15-5.16c.2-.2.5-.2.7 0Z"
                fill="currentColor"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Clear selection"
          class="fui-Dropdown__clearButton"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class=""
            fill="currentColor"
            height="1em"
            viewBox="0 0 20 20"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m4.09 4.22.06-.07a.5.5 0 0 1 .63-.06l.07.06L10 9.29l5.15-5.14a.5.5 0 0 1 .63-.06l.07.06c.18.17.2.44.06.63l-.06.07L10.71 10l5.14 5.15c.18.17.2.44.06.63l-.06.07a.5.5 0 0 1-.63.06l-.07-.06L10 10.71l-5.15 5.14a.5.5 0 0 1-.63.06l-.07-.06a.5.5 0 0 1-.06-.63l.06-.07L9.29 10 4.15 4.85a.5.5 0 0 1-.06-.63l.06-.07-.06.07Z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </label>
    <span
      class="fui-Textarea"
    >
      <textarea
        class="fui-Textarea__textarea"
        id="SourceInput-TextInput"
        rows="10"
        spellcheck="false"
        wrap="off"
      >
        csv text
      </textarea>
    </span>
    <br />
    <br />
    <br />
    <label
      class="fui-Label"
    >
      <span
        class="fui-Subtitle1 fui-Text"
      >
        Delimiter
      </span>
      <br />
      <div
        class="fui-Dropdown"
      >
        <button
          aria-expanded="false"
          class="fui-Dropdown__button"
          role="combobox"
          tabindex="0"
          type="button"
          value="Tab"
        >
          Tab
          <span
            class="fui-Dropdown__expandIcon"
          >
            <svg
              aria-hidden="true"
              class=""
              fill="currentColor"
              height="1em"
              viewBox="0 0 20 20"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.85 7.65c.2.2.2.5 0 .7l-5.46 5.49a.55.55 0 0 1-.78 0L4.15 8.35a.5.5 0 1 1 .7-.7L10 12.8l5.15-5.16c.2-.2.5-.2.7 0Z"
                fill="currentColor"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Clear selection"
          class="fui-Dropdown__clearButton"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class=""
            fill="currentColor"
            height="1em"
            viewBox="0 0 20 20"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m4.09 4.22.06-.07a.5.5 0 0 1 .63-.06l.07.06L10 9.29l5.15-5.14a.5.5 0 0 1 .63-.06l.07.06c.18.17.2.44.06.63l-.06.07L10.71 10l5.14 5.15c.18.17.2.44.06.63l-.06.07a.5.5 0 0 1-.63.06l-.07-.06L10 10.71l-5.15 5.14a.5.5 0 0 1-.63.06l-.07-.06a.5.5 0 0 1-.06-.63l.06-.07L9.29 10 4.15 4.85a.5.5 0 0 1-.06-.63l.06-.07-.06.07Z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </label>
    <br />
    <br />
    <label
      class="fui-Label"
    >
      <span
        class="fui-Subtitle1 fui-Text"
      >
        Newline sequence
      </span>
      <br />
      <div
        class="fui-Dropdown"
      >
        <button
          aria-expanded="false"
          class="fui-Dropdown__button"
          role="combobox"
          tabindex="0"
          type="button"
          value="LF"
        >
          LF
          <span
            class="fui-Dropdown__expandIcon"
          >
            <svg
              aria-hidden="true"
              class=""
              fill="currentColor"
              height="1em"
              viewBox="0 0 20 20"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.85 7.65c.2.2.2.5 0 .7l-5.46 5.49a.55.55 0 0 1-.78 0L4.15 8.35a.5.5 0 1 1 .7-.7L10 12.8l5.15-5.16c.2-.2.5-.2.7 0Z"
                fill="currentColor"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Clear selection"
          class="fui-Dropdown__clearButton"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class=""
            fill="currentColor"
            height="1em"
            viewBox="0 0 20 20"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m4.09 4.22.06-.07a.5.5 0 0 1 .63-.06l.07.06L10 9.29l5.15-5.14a.5.5 0 0 1 .63-.06l.07.06c.18.17.2.44.06.63l-.06.07L10.71 10l5.14 5.15c.18.17.2.44.06.63l-.06.07a.5.5 0 0 1-.63.06l-.07-.06L10 10.71l-5.15 5.14a.5.5 0 0 1-.63.06l-.07-.06a.5.5 0 0 1-.06-.63l.06-.07L9.29 10 4.15 4.85a.5.5 0 0 1-.06-.63l.06-.07-.06.07Z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </label>
    <br />
    <br />
    <button
      aria-label="Import CSV"
      class="fui-Button"
      type="button"
    >
      Import CSV
    </button>
    <br />
    <span
      class="fui-Text"
    >
      Â 
    </span>
    <div
      class=""
      style="margin-top: 30px;"
    >
      <span
        class="fui-Text"
      >
        <a
          href="/excel-csv-import/?page=about"
        >
          <button
            class="fui-Link"
            type="button"
          >
            About
          </button>
        </a>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
